### PR TITLE
Implement git-reset --merge and --keep options

### LINKED ISSUE #7358: Add "git reset --keep" to reset dialog
## Feature description

When I reset a branch in Git Extensions, I can pick soft/mixed/hard, but not keep.

`git reset --keep` moves HEAD and applies changes between old and new HEAD. It also keeps changes in the working directory. (IDK how the index comes into play, but that behavior differs between --merge and --keep. Documentation at https://git-scm.com/docs/git-reset)

"keep" is useful because it operates rather like switching branches (switch HEAD and update the filesystem to match), and is nondestructive (will not erase uncommitted changes in the working directory). If file A is modified locally and file B is different between the old/new HEAD, only file B will be modified (in both `git checkout` and `git reset --keep`). If file A is modified locally and is different between old/new HEAD, the reset is aborted.

Jetbrains IDEs already exposes an option to "Keep" when performing a git reset.

## Environment

- Git Extensions 3.2.1.6628
- Build bc07663c3c893f1d896520e0bdcfbe35cfe4b729
- Git 2.23.0.windows.1
- Microsoft Windows NT 10.0.18362.0
- .NET Framework 4.8.4018.0
- DPI 120dpi (125% scaling)

### PR DESCRIPTION
**Pull request by R0nd**
Fixes #7358

## Proposed changes

- Add git-reset --merge and --keep options to the reset dialog

### Before

image

### After

image

## Test methodology 

- Added appropriate test cases to GitCommandHelpersTest
- Manually tested the new options and confirmed they generate the expected commands

## Test environment(s) 

- GIT 2.14.1
- Windows 10

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by R0nd** [comment]
I'd love some advice on the wording of option descriptions and background colors of the new radios.

**Comment by mstv** [comment]
I'd sort the radiobuttons according to the severity of their effect, i.e. `hard` should be the last one.
Colors: Not sure, yellow or orange.
Labels: Explanation in consistent sequence, please, e.g. working directory first, index second.
I'd avoid mixing "index" and "[un]staged" or always mention both. E.g. the quite technical Git term "[commit] index" is not used in the FormCommit but "unstaged" and "staged". (This consideration also applies to the virtual commit "Commit index" in FormBrowse.)

**Comment by gerhardol** [comment]
I agree mostly with mstv
For me, the effects of the options are
soft - mixed - keep - merge - keep
keep, merge both yellow
Add a link to Git documentation
Texts to be updated...

Git uses officially Working Tree and [Commit] Index.
There has been some discussion around this about 1.5 years ago, staged/unstaged was preferred by the community, but I do not believe that GE should differ from the official Git terminology (even if even I do not believe that the Git naming is good here). Stage/unstage is is used as the action (verb) to move files though, I believe it is OK in FormCommit.
The virtual commits in Browse is tuned to "Working directory" and consitently use "Commit index", I suggest the same here.

**Comment by R0nd** [comment]
- Reordered options
- Reworded labels
- Set new options' background colors to "warning-yellow"
- Added link to git-reset documentation

File: GitCommands/ArgumentBuilderExtensions.cs [hunk 1]
```diff
@@ -226,6 +226,10 @@ string GetArgument()
                         return "--mixed";
                     case ResetMode.Hard:
                         return "--hard";
+                    case ResetMode.Merge:
```
Comment (gerhardol) [inline GitCommands/ArgumentBuilderExtensions.cs]: I suggest the selected order is the same in the code as in the GUI - dont change this yet though, I am not sure of the order myself...

File: GitUI/HelperDialogs/FormResetCurrentBranch.cs [hunk 2]
```diff
@@ -15,7 +16,9 @@ public enum ResetType
         {
             Soft,
             Mixed,
-            Hard
+            Hard,
+            Merge,
+            Keep,
```
Comment (gerhardol) [inline GitUI/HelperDialogs/FormResetCurrentBranch.cs]: comma last in the list

**Comment by RussKie** [comment]
I almost feel we need to convert the radio button list to a dropdown list as it has grown dramatically.
It also kind of feels intimidating presenting so may different options to a user at once.

What do you think?

Windows UI guidelines
* https://docs.microsoft.com/en-us/windows/win32/uxguide/ctrl-radio-buttons
* https://docs.microsoft.com/en-us/windows/win32/uxguide/ctrl-drop

**Comment by RussKie** [comment]
> * Reordered options
> * Reworded labels
> * Set new options' background colors to "warning-yellow"
> * Added link to git-reset documentation

Please update the screenshot

**Comment by pmiossec** [comment]
I also prefer keeping the radio buttons because that permits to compare the associated explanations (because that begins to become difficult...)

File: GitUI/HelperDialogs/FormResetCurrentBranch.Designer.cs [hunk 3]
```diff
@@ -180,49 +220,62 @@ private void InitializeComponent()
...
             // 
+            // linkGitReset
+            // 
+            this.linkGitReset.AutoSize = true;
+            this.linkGitReset.Location = new System.Drawing.Point(6, 417);
+            this.linkGitReset.Name = "linkGitReset";
+            this.linkGitReset.Size = new System.Drawing.Size(91, 13);
+            this.linkGitReset.TabIndex = 4;
+            this.linkGitReset.TabStop = true;
+            this.linkGitReset.Text = "More see git-reset";
```
Comment (pmiossec) [inline GitUI/HelperDialogs/FormResetCurrentBranch.Designer.cs]: I don't really like this label. I would prefer "For more information, see 'git reset' documentation"

File: GitUI/HelperDialogs/FormResetCurrentBranch.cs [hunk 4]
```diff
@@ -98,5 +115,10 @@ private void Cancel_Click(object sender, EventArgs e)
             DialogResult = DialogResult.Cancel;
             Close();
         }
+
+        private void linkGitReset_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
+        {
+            Process.Start(@"https://git-scm.com/docs/git-reset");
```
Comment (pmiossec) [inline GitUI/HelperDialogs/FormResetCurrentBranch.cs]: Perhaps you could link directly to the interesting section? https://git-scm.com/docs/git-reset#Documentation/git-reset.txt-emgitresetemltmodegtltcommitgt

**Comment by R0nd** [comment]
Updated the documentation link

**Comment by gerhardol** [comment]
> I almost feel we need to convert the radio button list to a dropdown list as it has grown dramatically.
It also kind of feels intimidating presenting so may different options to a user at once.
What do you think?

I leaning on 'advanced' expansion
Can accept radioboxes

**Comment by R0nd** [comment]
> 
> 
> Sorry, I don't like the added link.
> Can we use ["Help" button](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.form.helpbutton?view=netframework-4.8) instead? See: [HelpRequested event](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.control.helprequested?view=netframework-4.8)

To be honest I copy-pasted it from another dialog (Go to commit) because I wanted it to be consistent across the UI and AFAIK there's not many links like this currently.

I agree with changing it to a help button, but I think this prompts a conversation (outside the scope of this task) about standardizing help buttons.

**Comment by RussKie** [comment]
Yes, agree, we have a lot of inconsistency in UI. There is an open issue
that - #6183.
However we need to start somewhere, we may as well do it now.

**Comment by R0nd** [comment]
Added a help button that opens a link to the git-reset documentation on currently selected reset mode.

**Comment by mstv** [comment]
Update the screenshot, please, in order to ease reviews.

**Comment by R0nd** [comment]
> 
> 
> Update the screenshot, please, in order to ease reviews.

The screenshot should already be current

**Comment by mstv** [comment]
> The screenshot should already be current

Thank you, yes, it is. I didn't notice the help button in the title bar.

**Comment by pmiossec** [comment]
>Thank you, yes, it is. I didn't notice the help button in the title bar.

That's what worries me...

Even if it seems to be a standard pattern, I never used it and I don't know if i am representative of all the user and if users will see and use it 🤔

File: GitUI/HelperDialogs/FormResetCurrentBranch.Designer.cs [hunk 5]
```diff
@@ -182,49 +198,75 @@ private void InitializeComponent()
...
             this.flowLayoutPanel1.Name = "flowLayoutPanel1";
-            this.flowLayoutPanel1.Size = new System.Drawing.Size(689, 49);
+            this.flowLayoutPanel1.Size = new System.Drawing.Size(461, 29);
             this.flowLayoutPanel1.TabIndex = 3;
             // 
             // FormResetCurrentBranch
             // 
             this.AcceptButton = this.Ok;
-            this.AutoScaleDimensions = new System.Drawing.SizeF(144F, 144F);
+            this.AutoScaleDimensions = new System.Drawing.SizeF(96F, 96F);
```
Comment (RussKie) [inline GitUI/HelperDialogs/FormResetCurrentBranch.Designer.cs]: 👍

**Comment by RussKie** [comment]
Please update the translations

> Please update English translations and re-submit the pull-request.
For more information please refer to https://github.com/gitextensions/gitextensions/wiki/Translations

**Comment by R0nd** [comment]
> 
> 
> Please update the translations
> 
> > Please update English translations and re-submit the pull-request.
> > For more information please refer to https://github.com/gitextensions/gitextensions/wiki/Translations

Sorry, for some reason the update script isn't very reliable on my system. Had to delete the translation files and regenerate them.

**Comment by RussKie** [comment]
Thank you

**Comment by jimbo1qaz** [comment]
Thanks @R0nd for implementing this!

**Pull request merged**